File: trabajo 13/hash.py
class ShellSort:
    # Muestra los elementos del arreglo en una sola línea
    @staticmethod
    def displayArr(inputArr):
        for k in inputArr:
            print(k, end=" ")
        print()

    # Ordena un arreglo usando el algoritmo Shell Sort
    def sort(self, inputArr):
        size = len(inputArr)
        gapSize = size // 2  # Se empieza con un "salto" grande

        # Mientras el salto sea mayor que cero, seguimos ordenando
        while gapSize > 0:
            # Recorremos los elementos desde 'gapSize' hasta el final
            for j in range(gapSize, size):
                val = inputArr[j]  # Elemento actual a comparar
                k = j
                # Compara y mueve los elementos que estén fuera de orden
                while k >= gapSize and inputArr[k - gapSize] > val:
                    inputArr[k] = inputArr[k - gapSize]
                    k = k - gapSize
                inputArr[k] = val  # Inserta el elemento en su posición correcta

            gapSize = gapSize // 2  # Reduce el salto a la mitad en cada iteración

        return 0


if __name__ == "__main__":
    inputArr = [36, 38, 44, 11, 19, 30, 28, 50]

    print("Arreglo desordenado:")
    ShellSort.displayArr(inputArr)

    obj = ShellSort()
    obj.sort(inputArr)

    print("Arreglo ordenado:")
    ShellSort.displayArr(inputArr)

File: trabajo 13/hash.js
class ShellSort {
    // Muestra los elementos del arreglo en una sola línea
    static mostrarArreglo(arreglo) {
        console.log(arreglo.join(" "));
    }

    // Ordena un arreglo usando el algoritmo Shell Sort
    ordenar(arreglo) {
        let tamaño = arreglo.length;
        let salto = Math.floor(tamaño / 2); // Se empieza con un "salto" grande

        // Mientras el salto sea mayor que cero, seguimos ordenando
        while (salto > 0) {
            // Recorremos los elementos desde 'salto' hasta el final
            for (let j = salto; j < tamaño; j++) {
                let valor = arreglo[j]; // Elemento actual a comparar
                let k = j;

                // Compara y mueve los elementos que estén fuera de orden
                while (k >= salto && arreglo[k - salto] > valor) {
                    arreglo[k] = arreglo[k - salto];
                    k = k - salto;
                }

                // Inserta el elemento en su posición correcta
                arreglo[k] = valor;
            }

            // Reduce el salto a la mitad en cada iteración
            salto = Math.floor(salto / 2);
        }
    }
}


function main() {
    let arreglo = [36, 38, 44, 11, 19, 30, 28, 50];

    console.log("Arreglo desordenado:");
    ShellSort.mostrarArreglo(arreglo);

    const obj = new ShellSort();
    obj.ordenar(arreglo);

    console.log("Arreglo ordenado:");
    ShellSort.mostrarArreglo(arreglo);
}

main();
